### PR TITLE
[RayJob][Status][6/n] Redefine JobDeploymentStatusComplete and clean up K8s Job after TTL

### LINKED ISSUE #1233: [Bug] RayJob does not enter `Complete` state after job application failure
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/kuberay/issues) and found no similar issues.

### KubeRay Component

ray-operator

### What happened + What you expected to happen

Start the kuberay operator, and then  run `kubectl apply -f /Users/archit/kuberay/ray-operator/config/samples/ray_v1alpha1_rayjob.yaml` after modifying this YAML to include a typo or some other bug in the `entrypoint` field (to cause an application error.

Then `kubectl describe rayjob rayjob-sample` will show

```
  Job Deployment Status:  Running
  Job Id:                 rayjob-sample-prbts
  Job Status:             FAILED
  Message:                Job failed due to an application error, last available logs (truncated to 20,000 chars):
python: can't open file '/home/ray/samples/sample_code.ppy': [Errno 2] No such file or directory
```

The expected behavior is the status is Complete, not Running.

Likely related, the kuberay operator log prints the following in a loop every 3 seconds:

```
2023-07-11T17:58:08.015Z        INFO    controllers.RayJob      reconciling RayJob  {"NamespacedName": "default/rayjob-sample"}
2023-07-11T17:58:08.015Z        INFO    controllers.RayJob      Found associated RayCluster for RayJob       {"rayjob": "rayjob-sample", "raycluster": "default/rayjob-sample-raycluster-2h7ds"}
2023-07-11T17:58:08.016Z        INFO    controllers.RayJob      K8s job successfully retrieved       {"RayJob": "rayjob-sample", "jobId": "rayjob-sample"}
```

### Reproduction script

Above

### Anything else

Happens every time

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by kevin85421**
## Why are these changes needed?

* Without this PR, the transition between `Running` and `Complete` is pretty confusing. If `ShutdownAfterJobFinishes` is false, the RayJob's `JobDeploymentStatus` will be `JobDeploymentStatusRunning` although the Ray job is `Succeeded` or `Failed`. Both `Succeeded` and `Failed` are terminal states, so the job can't have any state transition. However, KubeRay still has to process a lot of unnecessary code, including sending a request to the RayCluster to check the status of the Ray job.

* This PR redefines `JobDeploymentStatusComplete`. `JobDeploymentStatusComplete` means that the JobStatus is `Succeeded` or `Failed`, and we only check TTL when the RayJob is `JobDeploymentStatusComplete`.

* Currently, the lifecycles of the submitter K8s Job and its Pod are not well-defined. For example, the K8s Job will not be deleted any matter TTL and `suspend`. In addition, the default deletion behavior for a Kubernetes Job is `orphanDependents`, so the Pod will not be deleted cascadingly after the Kubernetes Job is deleted. In this PR, we set TTL for the K8s Job the same as `RayJob.Spec.TTLSecondsAfterFinished`. I will chat with users to finalize the behavior in the follow up PRs

## Related issue number

Closes #1233 

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [x] Unit tests
   - [x] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: ray-operator/controllers/ray/rayjob_controller.go [hunk 1]
```diff
@@ -135,28 +158,10 @@ func (r *RayJobReconciler) Reconcile(ctx context.Context, request ctrl.Request)
 	// include STOPPED which is also a terminal status because `suspend` requires to stop the Ray job gracefully before
 	// delete the RayCluster.
 	if isJobSucceedOrFail(rayJobInstance.Status.JobStatus) {
-		// If the function `updateState` updates the JobStatus to Complete successfully, we can skip the reconciliation.
```
Comment (kevin85421) [inline ray-operator/controllers/ray/rayjob_controller.go:138]: With this PR, we only delete RayCluster (TTL) if the `JobDeploymentStatus` is `JobDeploymentStatusComplete`. Hence, we don't need to check whether it is deleted or not if the status if not `JobDeploymentStatusComplete`.

File: ray-operator/controllers/ray/rayjob_controller.go [hunk 2]
```diff
@@ -268,39 +272,7 @@ func (r *RayJobReconciler) Reconcile(ctx context.Context, request ctrl.Request)
 		}
 	}
 
-	// Let's use rayJobInstance.Status.JobStatus to make sure we only delete cluster after the CR is updated.
```
Comment (kevin85421) [inline ray-operator/controllers/ray/rayjob_controller.go:271]: Move the logic to L125 with this PR.

File: ray-operator/controllers/ray/rayjob_controller.go [hunk 3]
```diff
@@ -268,39 +272,7 @@ func (r *RayJobReconciler) Reconcile(ctx context.Context, request ctrl.Request)
...
-	// TODO (kevin85421): Use the source of truth `jobInfo.JobStatus` instead.
-	if isJobPendingOrRunning(jobInfo.JobStatus) {
-		// Requeue the RayJob to poll its status from the running Ray job
-		r.Log.Info("Requeue the RayJob because the Ray job is not in a terminal state", "RayJob", rayJobInstance.Name, "JobStatus", jobInfo.JobStatus)
-		return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, nil
-	}
-	// Otherwise only reconcile the RayJob upon new events for watched resources
-	// to avoid infinite reconciliation.
-	return ctrl.Result{}, nil
+	return ctrl.Result{RequeueAfter: RayJobDefaultRequeueDuration}, nil
```
Comment (kevin85421) [inline ray-operator/controllers/ray/rayjob_controller.go:275]: The only case that we don't requeue the CR is when it is in `JobDeploymentStatusComplete` and TTL.

File: ray-operator/controllers/ray/rayjob_controller.go [hunk 4]
```diff
@@ -388,6 +360,12 @@ func (r *RayJobReconciler) createNewK8sJob(ctx context.Context, rayJobInstance *
 		},
 	}
 
+	// Without TTLSecondsAfterFinished, the job has a default deletion policy of `orphanDependents` causing
```
Comment (kevin85421) [inline ray-operator/controllers/ray/rayjob_controller.go:363]: https://kubernetes.io/docs/concepts/workloads/controllers/job/#ttl-mechanism-for-finished-jobs

**Comment by kevin85421** [comment]
By the way, I want to give a shout-out to @astefanutti. This is my first time running and updating RayJob e2e tests by myself. The process is quite user-friendly, and writing new tests is pretty straightforward!

**Review by astefanutti — APPROVED** [review]

**Comment by astefanutti** [comment]
> By the way, I want to give a shout-out to @astefanutti. This is my first time running and updating RayJob e2e tests by myself. The process is quite user-friendly, and writing new tests is pretty straightforward!

@kevin85421 Thanks a lot for the feedback, really appreciated! That's also a great encouragement for us to contribute more to the e2e tests 😃!

**Review by architkulkarni — APPROVED** [review]
Looks good to me!

**Pull request merged**